File: app/src/main/java/com/freetime/geoweather/WeatherIconMapper.java
package com.freetime.geoweather;

import android.os.Build;
import android.text.TextUtils;

import com.freetime.geoweather.R;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;

public class WeatherIconMapper {

    private static ZonedDateTime sunriseTime;
    private static ZonedDateTime sunsetTime;

    public static void setSunTimes(String sunriseIso, String sunsetIso) {
        try {
            if (!TextUtils.isEmpty(sunriseIso)) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    sunriseTime = ZonedDateTime.parse(sunriseIso + ":00Z")
                            .withZoneSameInstant(ZoneId.systemDefault());
                }
            }
            if (!TextUtils.isEmpty(sunsetIso)) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    sunsetTime = ZonedDateTime.parse(sunsetIso + ":00Z")
                            .withZoneSameInstant(ZoneId.systemDefault());
                }
            }
        } catch (Exception e) {
            sunriseTime = null;
            sunsetTime = null;
        }
    }

    private static boolean isDaytime() {
        if (sunriseTime == null || sunsetTime == null) return true;
        ZonedDateTime now = null;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            now = ZonedDateTime.now(ZoneId.systemDefault());
        }
        return now.isAfter(sunriseTime) && now.isBefore(sunsetTime);
    }

    public static int getWeatherIcon(int code) {
        boolean night = isDaytime();

        switch (code) {

            case 0:
                return night ? R.drawable.google_clear_night : R.drawable.google_clear_day;

            case 1:
                return night ? R.drawable.google_mostly_clear_night : R.drawable.google_mostly_clear_day;

            case 2:
                return night ? R.drawable.google_partly_cloudy_night : R.drawable.google_partly_cloudy_day;

            case 3:
                return night ? R.drawable.google_cloudy : R.drawable.google_cloudy;

            case 45:
            case 48:
                return night ? R.drawable.google_fog : R.drawable.google_fog;

            case 51:
            case 53:
            case 55:
                return night ? R.drawable.google_drizzle : R.drawable.google_drizzle;

            case 61:
            case 63:
            case 65:
                return night ? R.drawable.google_rain_with_sunny_dark : R.drawable.google_rain_with_sunny_light;

            case 71:
            case 73:
            case 75:
                return night ? R.drawable.google_snow_with_sunny_dark : R.drawable.google_snow_with_sunny_light;

            default:
                return night ? R.drawable.google_cloudy_with_sunny_dark : R.drawable.google_cloudy_with_sunny_light;
        }
    }
}

File: app/src/main/java/com/freetime/geoweather/WeatherIconMapper.kt
package com.freetime.geoweather

import androidx.annotation.DrawableRes
import com.freetime.geoweather.R as Res
import kotlinx.datetime.*

object WeatherIconMapper {
    private var sunriseTime: LocalDateTime? = null
    private var sunsetTime: LocalDateTime? = null

    fun setSunTimes(sunrise: String, sunset: String) {
        try {
            sunriseTime = Instant.parse(sunrise).toLocalDateTime(TimeZone.currentSystemDefault())
            sunsetTime = Instant.parse(sunset).toLocalDateTime(TimeZone.currentSystemDefault())
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun isDaytime(): Boolean {
        val sunrise = sunriseTime ?: return true
        val sunset = sunsetTime ?: return true
        val now = Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault())
        return now > sunrise && now < sunset
    }

    @DrawableRes
    fun getWeatherIcon(code: Int, theme: String = "google"): Int =
        getWeatherIcon(code, isDaytime(), theme)

    @DrawableRes
    fun getWeatherIcon(code: Int, isDay: Boolean, theme: String = "google"): Int {
        return when (code) {
            0 -> if (isDay) Res.drawable.google_clear_day else Res.drawable.google_clear_night
            1 -> if (isDay) Res.drawable.google_mostly_clear_day else Res.drawable.google_mostly_clear_night
            2 -> if (isDay) Res.drawable.google_partly_cloudy_day else Res.drawable.google_partly_cloudy_night
            3 -> Res.drawable.google_cloudy
            45, 48 -> Res.drawable.google_fog
            51, 53, 55 -> Res.drawable.google_drizzle
            56, 57 -> Res.drawable.google_drizzle
            61, 63, 65, 66, 67, 80, 81, 82 ->
                if (isDay) Res.drawable.google_rain_with_sunny_light else Res.drawable.google_rain_with_sunny_dark
            71, 73, 75, 77, 85, 86 ->
                if (isDay) Res.drawable.google_snow_with_sunny_light else Res.drawable.google_snow_with_sunny_dark
            95, 96, 99 ->
                if (isDay) Res.drawable.google_rain_with_sunny_light else Res.drawable.google_rain_with_sunny_dark
            else -> if (isDay) Res.drawable.google_cloudy_with_sunny_light else Res.drawable.google_cloudy_with_sunny_dark
        }
    }
}
